pkg/rbac/authorization.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rbac
+
+import (
+	"fmt"
+
+	"strings"
+
+	"github.com/dgrijalva/jwt-go"
+	api "github.com/onosproject/onos-lib-go/api/rbac"
+	"google.golang.org/grpc"
+)
+
+const (
+	// GroupsKey groups key in the token claims
+	GroupsKey = "groups"
+)
+
+// Rbac role based access control interface
+type Rbac interface {
+	Authorize() error
+}
+
+// NewAuthorization creates a new Authorization instance
+func NewAuthorization(claims jwt.MapClaims, info *grpc.UnaryServerInfo) Authorization {
+	return Authorization{
+		claims: claims,
+		info:   info,
+	}
+}
+
+// SetClaims set jwt token claims
+func (a *Authorization) SetClaims(claims jwt.MapClaims) {
+	a.claims = claims
+}
+
+// SetUnaryServerInfo sets unary server info
+func (a *Authorization) SetUnaryServerInfo(info *grpc.UnaryServerInfo) {
+	a.info = info
+}
+
+// Authorization authorization claimed information
+type Authorization struct {
+	claims jwt.MapClaims
+	info   *grpc.UnaryServerInfo
+}
+
+func verifyRules(rules []*api.Rule, fullMethod string) error {
+	// Retrieve service information and rpc method name
+	reqService, reqVerb := getMethodInformation(fullMethod)
+	for _, rule := range rules {
+		for _, service := range rule.Services {
+			if match(strings.ToLower(service), strings.ToLower(reqService)) {
+				for _, verb := range rule.Verbs {
+					if match(strings.ToLower(verb), reqVerb) {
+						return nil
+					}
+				}
+			}
+		}
+	}
+
+	return fmt.Errorf("no rule found to authorize the user")
+}
+
+func extractClaimedGroups(claims jwt.MapClaims) ([]string, error) {
+	var claimedGroups []interface{}
+	for key := range claims {
+		// extract claimed groups from the token
+		if key == GroupsKey {
+			claimedGroups = claims[GroupsKey].([]interface{})
+		}
+	}
+
+	// If the user does not claim any groups then we cannot authorize the user
+	if claimedGroups == nil {
+		return nil, fmt.Errorf("groups claim cannot be empty")
```

Comment:
at least give the `email` from the claim in the log message

Reply:
I can return the claims which shows the groups are not populated.

Reply:
Maybe, but at least the `email` is needed to know what user, it's referring to

Reply:
Will add it.